File: notes_app_react/services/note-service.js
import { ID } from "appwrite";
import databases, { listDocuments, APPWRITE_DATABASE_ID, APPWRITE_COLLECTION_ID } from "./database-service";

// Helper to normalize Appwrite document to local note shape
const normalize = (doc) => {
  return {
    id: doc.$id || doc.id,
    title: doc.title || "",
    content: doc.content || "",
    createdAt: doc.createdAt || doc.$createdAt || null,
    updatedAt: doc.updatedAt || doc.$updatedAt || null,
    userId: doc.userId || null,
    raw: doc,
  };
};

export const getNotes = async (userId = null) => {
  try {
    const queries = [];
    if (userId) {
      // Use Appwrite Query equal if userId is provided
      // Import Query dynamically to avoid errors if appwrite not installed yet
      // eslint-disable-next-line global-require
      const { Query } = require("appwrite");
      queries.push(Query.equal("userId", userId));
    }
    // Sort newest first using system createdAt field
    try {
      const { Query } = require("appwrite");
      queries.push(Query.orderDesc("$createdAt"));
    } catch (e) {
      // ignore if Query not available
    }

    const docs = await listDocuments(queries);
    return docs.map(normalize);
  } catch (error) {
    console.error("Error getting notes:", error);
    throw error;
  }
};

export const createNote = async (data) => {
  try {
    // Appwrite automatically manages $createdAt and $updatedAt
    const noteData = {
      title: data.title || "",
      content: data.content || "",
      userId: data.userId || "",
    };

    const response = await databases.createDocument(
      APPWRITE_DATABASE_ID,
      APPWRITE_COLLECTION_ID,
      ID.unique(),
      noteData
    );

    return normalize(response);
  } catch (error) {
    console.error("Error creating note:", error);
    throw error;
  }
};

export const deleteNote = async (noteId) => {
  try {
    await databases.deleteDocument(
      APPWRITE_DATABASE_ID,
      APPWRITE_COLLECTION_ID,
      noteId
    );
    return true;
  } catch (error) {
    console.error("Error deleting note:", error);
    throw error;
  }
};

export const updateNote = async (noteId, data) => {
  try {
    // Appwrite automatically updates $updatedAt
    const noteData = {
      title: data.title,
      content: data.content,
    };

    const response = await databases.updateDocument(
      APPWRITE_DATABASE_ID,
      APPWRITE_COLLECTION_ID,
      noteId,
      noteData
    );

    return normalize(response);
  } catch (error) {
    console.error("Error updating note:", error);
    throw error;
  }
};

File: notes_app/lib/services/note_service.dart
import 'package:appwrite/appwrite.dart';
import 'package:appwrite/models.dart';
import 'appwrite_config.dart';

class NoteService {
  late final Databases _databases;

  NoteService() {
    _databases = Databases(AppwriteConfig.client);
  }

  // Get all notes, potentially filtered by userId
  Future<List<Document>> getNotes({String? userId}) async {
    try {
      // Create query list - initially empty
      List<String> queries = [];

      // If userId is provided, add a filter
      if (userId != null) {
        queries.add(Query.equal('userId', userId));
      }

      // Add sorting by createdAt descending
      queries.add(Query.orderDesc('createdAt'));

      // Fetch documents from the database
      final response = await _databases.listDocuments(
        databaseId: AppwriteConfig.databaseId,
        collectionId: AppwriteConfig.collectionId,
        queries: queries,
      );

      return response.documents;
    } catch (e) {
      print('Error getting notes: $e');
      rethrow;
    }
  }

  // Create a new note
  Future<Document> createNote(Map<String, dynamic> data) async {
    try {
      // Add timestamps to the note data
      final noteData = {
        ...data,
        'createdAt': DateTime.now().toIso8601String(),
        'updatedAt': DateTime.now().toIso8601String(),
      };

      // Create a document in the database
      final response = await _databases.createDocument(
        databaseId: AppwriteConfig.databaseId,
        collectionId: AppwriteConfig.collectionId,
        documentId: ID.unique(),
        data: noteData,
      );

      return response;
    } catch (e) {
      print('Error creating note: $e');
      rethrow;
    }
  }

  // Delete a note by ID
  Future<void> deleteNote(String noteId) async {
    try {
      await _databases.deleteDocument(
        databaseId: AppwriteConfig.databaseId,
        collectionId: AppwriteConfig.collectionId,
        documentId: noteId,
      );
    } catch (e) {
      print('Error deleting note: $e');
      rethrow;
    }
  }

  // Update an existing note
  Future<Document> updateNote(String noteId, Map<String, dynamic> data) async {
    try {
      final noteData = {
        ...data,
        'updatedAt': DateTime.now().toIso8601String(),
      };

      final response = await _databases.updateDocument(
        databaseId: AppwriteConfig.databaseId,
        collectionId: AppwriteConfig.collectionId,
        documentId: noteId,
        data: noteData,
      );

      return response;
    } catch (e) {
      print('Error updating note: $e');
      rethrow;
    }
  }
}
